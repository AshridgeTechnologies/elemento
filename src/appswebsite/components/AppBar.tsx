import * as React from 'react';
import Mui_AppBar from '@mui/material/AppBar';
import Box from '@mui/material/Box';
import Toolbar from '@mui/material/Toolbar';
import Typography from '@mui/material/Typography';
import IconButton from '@mui/material/IconButton';
import MenuIcon from '@mui/icons-material/Menu';
import UserMenu from '../../shared/UserMenu'

export default function AppBar() {
    return (
        <Box sx={{ flexGrow: 1 }}>
            <Mui_AppBar position="sticky">
                <Toolbar variant="dense">
                    <IconButton
                        size="large"
                        edge="start"
                        color="inherit"
                        aria-label="menu"
                        sx={{ mr: 2 }}
                    >
                        <MenuIcon />
                    </IconButton>
                    <Typography variant="h6" component="div" sx={{ fontSize: 16, flexGrow: 1 }}>
                        Elemento Apps
                    </Typography>
                    <UserMenu/>
                </Toolbar>
            </Mui_AppBar>
        </Box>
    );
}