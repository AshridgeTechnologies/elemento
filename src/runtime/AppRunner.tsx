import React, {createRef, useEffect} from 'react'
import {StoreProvider} from './appData'
import {highlightElement} from './runtimeFunctions'
import {ErrorBoundary} from 'react-error-boundary'
import ErrorFallback from './ErrorFallback'

function SelectionProvider({children, onComponentSelected, selectedComponentId}: {children: React.ReactNode, onComponentSelected: (id: string) => void, selectedComponentId?: string}) {
    const containerRef = createRef()

    const selectionEventListener = (event: MouseEvent) => {
        if (event.altKey) {
            event.preventDefault()
            event.stopPropagation()
            onComponentSelected((event.target as HTMLElement).id)
        }
    }
    useEffect(() => (containerRef.current as HTMLElement).addEventListener('click', selectionEventListener) )
    useEffect(() => {
        if (selectedComponentId) {
            highlightElement(selectedComponentId)
        }
    })
    // @ts-ignore
    return <div ref={containerRef}>{children}</div>
}

type Properties = {appFunction: React.FunctionComponent<any>, onComponentSelected: (id: string) => void, selectedComponentId?: string}

export default function AppRunner({appFunction, onComponentSelected, selectedComponentId}: Properties) {
    return <ErrorBoundary FallbackComponent={ErrorFallback} resetKeys={[appFunction]}>
        <StoreProvider>
            <SelectionProvider onComponentSelected={onComponentSelected} selectedComponentId={selectedComponentId}>
                {React.createElement(appFunction)}
            </SelectionProvider>
        </StoreProvider>
    </ErrorBoundary>
}