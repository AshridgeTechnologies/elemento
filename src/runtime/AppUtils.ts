import {ensureSlash} from './runtimeFunctions'

export default class AppUtils {
    constructor(private resourceUrl: string | undefined ) {}

    private get prefix() {
        return this.resourceUrl ? this.resourceUrl.replace(/\/$/, '') : ''
    }
    getFullUrl(url: string | undefined) {
        if (!url) return url
        if (url.match(/^https?:\/\//)) return url
        return this.prefix + ensureSlash(url)
    }

}