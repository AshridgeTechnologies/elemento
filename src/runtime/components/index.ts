export {default as App} from './App'
export {default as Button} from './Button'
export {default as Icon} from './Icon'
export {default as Image} from './Image'
export {default as UserLogon} from './UserLogon'
export {default as Menu} from './Menu'
export {default as MenuItem} from './MenuItem'
export {default as NumberInput} from './NumberInput'
export {default as DateInput} from './DateInput'
export {default as TrueFalseInput} from './TrueFalseInput'
export {default as SpeechInput} from './SpeechInput'
export {default as Data} from './Data'
export {default as Calculation} from './Calculation'
export {default as Timer} from './Timer'
export {default as Collection} from './Collection'
export {default as Page} from './Page'
export {default as ComponentElement} from './ComponentElement'
export {default as Form} from './Form'
export {default as BaseFormState} from './FormState'
export {default as SelectInput} from './SelectInput'
export {default as TextElement} from './TextElement'
export {default as Layout} from './Layout'
export {default as Block} from './Block'
export {default as AppBar} from './AppBar'
export {default as ListElement} from './ListElement'
export {default as ListItem} from './ListItem'
export {default as ItemSet} from './ItemSet'
export {default as ItemSetItem} from './ItemSetItem'
export {default as TextInput} from './TextInput'
export {default as MemoryDataStore} from './MemoryDataStore'
export {default as FileDataStore} from './FileDataStore'
export {default as FileDataStoreImpl} from './FileDataStoreImpl'
export {default as IdbDataStoreImpl} from './IdbDataStoreImpl'
export {default as BrowserDataStore} from './BrowserDataStore'
export {default as FirestoreDataStore} from './FirestoreDataStore'
export {default as FirestoreDataStoreImpl} from './FirestoreDataStoreImpl'
export {default as ServerAppConnector} from './ServerAppConnector'
export {default as Adapter} from './Adapter'
