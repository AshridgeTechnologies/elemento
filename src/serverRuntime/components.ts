export {default as FirestoreDataStore} from './FirestoreDataStore'
export {default as Collection} from './Collection'

// NOTE: Must add the name of each component to ./names.ts