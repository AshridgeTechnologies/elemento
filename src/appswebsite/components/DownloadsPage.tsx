import React from 'react'
import {Box, Container, Link, Stack, Typography} from '@mui/material'
import AppBar from '../../shared/AppBar'
import {theme} from '../../shared/styling'
import {ThemeProvider} from '@mui/material/styles'

export default function LearnPage() {

    return <ThemeProvider theme={theme}>
        <Box display='flex' flexDirection='column' height='100%' width='100%'>
            <Box flex='0'>
                <AppBar title='Elemento Online - Downloads' userMenu={false}/>
            </Box>
            <Box flex='1' minHeight={0}>
                <Container>
                    <Stack maxWidth={600} spacing={3} mt={3}>
                        <Typography variant='h4'>Elemento Development Server</Typography>
                        <Typography>Please download the correct version for your operating system.</Typography>
                        <Typography>Your computer will complain that the app is not known, approved, signed etc.<br/>
                            We do not currently have the resources to jump through all the hoops required to get the app
                            approved on each operating system,
                            so you will just have to trust us here, and continue past all the warnings.
                        </Typography>
                        <Link href='/devServer/win/ElementoDevServer.exe'><Typography variant='h5'>Windows</Typography></Link>
                        <Link href='/devServer/macos/ElementoDevServer.zip'><Typography variant='h5'>Mac</Typography></Link>
                        <Link href='/devServer/linux/ElementoDevServer'><Typography variant='h5'>Linux</Typography></Link>
                    </Stack>
                </Container>
            </Box>
        </Box>
    </ThemeProvider>
}
