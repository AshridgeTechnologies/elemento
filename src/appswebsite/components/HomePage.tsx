import React from 'react'
import {Box, Container, Link, Stack, Typography} from '@mui/material'
import AppBar from '../../shared/AppBar'
import {theme} from '../../shared/styling'
import {ThemeProvider} from '@mui/material/styles'
import HomeIcon from '@mui/icons-material/Home'
import LightbulbIcon from '@mui/icons-material/LightbulbTwoTone'
import CreateIcon from '@mui/icons-material/CreateTwoTone'
import AodIcon from '@mui/icons-material/AodTwoTone'
import ArrowCircleRightTwoToneIcon from '@mui/icons-material/ArrowCircleRightTwoTone';

const primary = theme.palette.primary.main
const secondary = theme.palette.secondary.main

function Logo() {
    return <Box color={secondary} border='8px solid' borderRadius={4} width={108} height={104} paddingTop='10px' paddingX='3px'>
        <Stack>
            <Typography fontSize='100px' fontFamily='"Helvetica"' align='center' lineHeight={0.8}>El</Typography>
            <Typography fontSize='16px' lineHeight={1}>42</Typography>
        </Stack>
    </Box>
}

function SectionLink(props: {title: string, href: string, icon: any, children: any}) {
    const Icon = props.icon
    return <Box border='4px solid' borderColor={primary} borderRadius={4} width='300px' paddingY='10px' paddingX='16px'>
        <Link href={props.href} underline='none'>
            <Stack spacing={2}>
                <Box textAlign='center'><Icon sx={{ fontSize: 180 }} /></Box>
                <Link component='div' underline='hover' href={props.href} variant='h5' textAlign='center'>{props.title}</Link>
                <Typography fontSize='20px' color='#555'>{props.children}</Typography>
                <Box textAlign='center'><ArrowCircleRightTwoToneIcon sx={{ fontSize: 32, color: secondary }} /></Box>
            </Stack>
        </Link>
    </Box>
}

export default function HomePage() {

    return <ThemeProvider theme={theme}>
        <Box display='flex' flexDirection='column' height='100%' width='100%'>
            <Box flex='0'>
                <AppBar title='Elemento Apps - Home'/>
            </Box>
            <Box flex='1' minHeight={0} mt={2}>
                <Container maxWidth="lg">
                    <Stack spacing={2}>
                        <Stack direction='row'>
                            <Logo/>
                            <Stack ml={4}>
                                <Typography fontWeight='bold' variant='h3'  color={secondary} align='center'>Elemento</Typography>
                                <Typography variant='h4' color={secondary} align='center' mt={2}>
                                    Everything you need to make and use your own software
                                </Typography>
                            </Stack>
                        </Stack>
                        <Stack direction='row' justifyContent='space-between' alignItems='stretch' marginTop='30px'>
                            <SectionLink title='Learn' href='/learn' icon={LightbulbIcon}>
                                A great range of resources to teach you how to get started and create your first app with Elemento.
                            </SectionLink>
                            <SectionLink title='Create' href='/studio' icon={CreateIcon}>
                                Put your app together and preview it live in Elemento Studio
                            </SectionLink>
                            <SectionLink title='Use' href='/run' icon={AodIcon}>
                                Run your app - or go back to the apps you have already made.
                            </SectionLink>
                        </Stack>
                    </Stack>
                </Container>
            </Box>
        </Box>
    </ThemeProvider>
}
