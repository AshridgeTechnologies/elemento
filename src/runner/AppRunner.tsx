import React, {createRef, useEffect} from 'react'
import {AppStoreHook, StoreProvider} from '../runtime/appData'
import {highlightElement} from '../runtime/runtimeFunctions'
import {ErrorBoundary} from 'react-error-boundary'
import ErrorFallback from './ErrorFallback'
import AppContext from '../runtime/AppContext'
import AppUtils from '../runtime/AppUtils'

export const AppUtilsContext = React.createContext<AppUtils | null>(null)

function SelectionProvider({children, onComponentSelected, selectedComponentId}: {children: React.ReactNode, onComponentSelected: (id: string) => void, selectedComponentId?: string}) {
    const containerRef = createRef()

    const selectionEventListener = (event: MouseEvent) => {
        if (event.altKey && onComponentSelected) {
            event.preventDefault()
            event.stopPropagation()
            const target = event.target as HTMLElement
            const id = target.id || target.closest('[id]')?.id
            if (id) {
                onComponentSelected(id)
            }
        }
    }
    useEffect(() => (containerRef.current as HTMLElement).addEventListener('click', selectionEventListener), [] )
    useEffect(() => highlightElement(selectedComponentId), [selectedComponentId])
    // @ts-ignore
    return <div id='selectionProvider' style={{height: '100%', width:'100%'}} ref={containerRef}>{children}</div>
}

type Properties = {appFunction: React.FunctionComponent<any>, appContext: AppContext, resourceUrl?: string,
    onComponentSelected: (id: string) => void, selectedComponentId?: string, appStoreHook?: AppStoreHook }

export default function AppRunner({appFunction, appContext, resourceUrl, onComponentSelected, selectedComponentId, appStoreHook}: Properties) {
    return <ErrorBoundary FallbackComponent={ErrorFallback} resetKeys={[appFunction]}>
        <StoreProvider appStoreHook={appStoreHook}>
            <AppUtilsContext.Provider value={new AppUtils(resourceUrl)}>
                <SelectionProvider onComponentSelected={onComponentSelected} selectedComponentId={selectedComponentId}>
                    {React.createElement(appFunction, {appContext})}
                </SelectionProvider>
            </AppUtilsContext.Provider>
        </StoreProvider>
    </ErrorBoundary>
}