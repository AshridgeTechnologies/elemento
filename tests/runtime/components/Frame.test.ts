/**
 * @jest-environment jsdom
 */

import {createElement} from 'react'
import {Frame} from '../../../src/runtime/components/index'
import {componentJSON, snapshot, valueObj, wait} from '../../testutil/testHelpers'
import {AppUtilsContext} from '../../../src/runner/AppRunner'
import AppUtils from '../../../src/runtime/AppUtils'

const inAppUtilsProvider = (prefix: string | undefined, el: any) => {
    const appUtils = new AppUtils(prefix)
    return createElement(AppUtilsContext.Provider, {value: appUtils}, el)
}

test('Frame element produces frame output with default properties',
    snapshot(inAppUtilsProvider(undefined, createElement(Frame, {path: 'app.page1.photo', source: 'sidebar.pdf', })))
)

test('Frame element not visible with display false',
    snapshot(inAppUtilsProvider('http://example.com/anApp/somewhere', createElement(Frame, {path: 'app.page1.photo', source: 'sidebar.pdf', show: false})))
)

test('Frame element produces output with properties supplied as state values', async () => {
        const element = inAppUtilsProvider('http://example.com/anApp/somewhere',
            createElement(Frame, {path: 'app.page1.photo', source: valueObj('sidebar.pdf'), show: valueObj(true),
                                                            styles: {width: 200, height: 100, marginBottom:'1em'} }))
        await wait(10)
        expect(componentJSON(element)).toMatchSnapshot()
    }
)


