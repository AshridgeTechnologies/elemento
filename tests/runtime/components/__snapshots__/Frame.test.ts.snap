// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Frame element not visible with display false 1`] = `
<iframe
  id="app.page1.photo"
  src="http://example.com/anApp/somewhere/sidebar.pdf"
  style={
    {
      "border": "none",
      "display": "none",
    }
  }
/>
`;

exports[`Frame element produces frame output with default properties 1`] = `
<iframe
  id="app.page1.photo"
  src="/sidebar.pdf"
  style={
    {
      "border": "none",
    }
  }
/>
`;

exports[`Frame element produces output with properties supplied as state values 1`] = `
<iframe
  id="app.page1.photo"
  src="http://example.com/anApp/somewhere/sidebar.pdf"
  style={
    {
      "border": "none",
      "height": "100px",
      "marginBottom": "1em",
      "width": "200px",
    }
  }
/>
`;
